File: src/lib/geo.ts
// Calculs géographiques.

const R = 6371000; // rayon terrestre en mètres

function toRad(deg: number): number {
  return (deg * Math.PI) / 180;
}

/** Distance en mètres entre deux coordonnées (formule de haversine). */
export function haversineMeters(
  a: { lat: number; lon: number },
  b: { lat: number; lon: number },
): number {
  const dLat = toRad(b.lat - a.lat);
  const dLon = toRad(b.lon - a.lon);
  const lat1 = toRad(a.lat);
  const lat2 = toRad(b.lat);
  const h =
    Math.sin(dLat / 2) ** 2 + Math.cos(lat1) * Math.cos(lat2) * Math.sin(dLon / 2) ** 2;
  return 2 * R * Math.asin(Math.min(1, Math.sqrt(h)));
}

/**
 * Décime un tracé en ne conservant qu'un point tous les `minMeters` (départ et
 * arrivée toujours préservés). Réduit le nombre de sommets dessinés sur une
 * longue sortie sans altérer la forme — purement pour l'affichage, la base
 * conserve la résolution pleine.
 */
export function decimateByDistance<T extends { lat: number; lon: number }>(
  points: T[],
  minMeters: number,
): T[] {
  if (points.length <= 2) return points;
  const out: T[] = [points[0]];
  let last = points[0];
  for (let i = 1; i < points.length - 1; i++) {
    if (haversineMeters(last, points[i]) >= minMeters) {
      out.push(points[i]);
      last = points[i];
    }
  }
  out.push(points[points.length - 1]);
  return out;
}

File: app/src/main/java/ovh/battistella/elan/domain/Geo.kt
// Calculs géographiques.
package ovh.battistella.elan.domain

import kotlin.math.PI
import kotlin.math.asin
import kotlin.math.cos
import kotlin.math.min
import kotlin.math.sin
import kotlin.math.sqrt

/** Rayon terrestre moyen en mètres. */
private const val EARTH_RADIUS_M = 6_371_000.0

/** Coordonnée nue, pour les appelants qui n'ont pas de point typé sous la main. */
data class LatLon(override val lat: Double, override val lon: Double) : GeoPoint

private fun toRad(deg: Double): Double = deg * PI / 180

/** Distance en mètres entre deux coordonnées (formule de haversine). */
fun haversineMeters(a: GeoPoint, b: GeoPoint): Double {
    val dLat = toRad(b.lat - a.lat)
    val dLon = toRad(b.lon - a.lon)
    val lat1 = toRad(a.lat)
    val lat2 = toRad(b.lat)
    val sinLat = sin(dLat / 2)
    val sinLon = sin(dLon / 2)
    val h = sinLat * sinLat + cos(lat1) * cos(lat2) * sinLon * sinLon
    return 2 * EARTH_RADIUS_M * asin(min(1.0, sqrt(h)))
}

/**
 * Décime un tracé en ne conservant qu'un point tous les `minMeters` (départ et
 * arrivée toujours préservés). Purement pour l'affichage : la base conserve la
 * résolution pleine.
 */
fun <T : GeoPoint> decimateByDistance(points: List<T>, minMeters: Double): List<T> {
    if (points.size <= 2) return points
    val out = ArrayList<T>()
    out.add(points[0])
    var last = points[0]
    for (i in 1 until points.size - 1) {
        if (haversineMeters(last, points[i]) >= minMeters) {
            out.add(points[i])
            last = points[i]
        }
    }
    out.add(points[points.size - 1])
    return out
}
